### ISSUE TITLE
Implement build caching support in esbuild-based browser application builder

### ISSUE THREAD

**Issue by eneajaho** [angular/angular-cli#25130]
### Command

build, serve

### Description

Implement build caching support in esbuild-based browser application builder

### Describe the solution you'd like

Just like the webpack builder supports build caching, it would be nice to have it also in esbuild builder. As currently the webpack builder is faster on cold builds after the first run compared to esbuild one. 

### Describe alternatives you've considered

Use webpack builder.

**Comment by vbraun** [comment]
@intellix I'm also getting slightly slower incremental rebuilds, you are not the only one. Though build caching (this ticket) should only affect cold starts.

**Comment by bjornharvold** [comment]
When using the new esbuild feature from Angular 17.0.3 and TailwindCss, it takes us 500+ seconds to start up a large application. When we make a simple code change, it takes 280+ seconds for esbuild to rebuild.

The esbuild build story is GREAT. Working with esbuild and TailwindCSS on a daily basis is a completely different story.

Wondering what's possible here. The current situation is not tenable.

FYI Before we used Nx's incremental webpack build and that worked in an "acceptable" way. If the change was in the "top most" part of the application, a rebuild was snappy. If the change was made in the "lower part" of the app, the dev server would have to work its way up to the top by compiling everything in between.

You definitely see a delay when having to pre-process the CSS in a larger application in comparison to an Angular app that runs with SCSS / Bootstrap for example.

Thoughts from the Team Angular highly requested 🍺

**Comment by vbraun** [comment]
We generate a static tailwind scss, that works fine (but unnused classes are then not pruned). I suggest you open a separate ticket as its specific to the built-in Angular tailwindcss support.

**Comment by kklocker** [comment]
Having the same issue here: After upgrading to angular 17 we want to try out the new application builder, but increasing the build times (compared to cached webpack builds) is a tough one to swallow

**Comment by JeanRemiDelteil** [comment]
Is the `@angular-devkit/build-angular:application` builder using cache at all ?

On our application we have no differences between the first build with no cache, and the following cold build with cache.

Here are some approximate numbers collected from multiple executions from a dev machine:

| Builder | Cold build, no cache | Cold build, with cache | Hot rebuild<br/>(small change in a leaf component) | Hot rebuild<br/>(small change in a root scss) |
|---------|----------------------|------------------------|----------------------------------------------------|-----------------------------------------------|
| Webpack | 202s                 | 60s                    | 17s                                                | 94s                                           |
| esBuild | 111s                 | 109s                   | 7s                                                 | 95s                                           |

In our case though there multiple goals:
- CI runs: Cold build, with cache
- Local UI dev: All of the above conditions
- Back-end devs: Cold build, with cache

Basically there are no real improvement to use the esBuild builder.
We even have costlier CI cycles (and CI cycle typically takes longer then the above metrics, as the machines are less powerful).
